File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/SnowflakeInsertDestinationAcceptanceTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.snowflake;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.airbyte.cdk.db.factory.DataSourceFactory;
import io.airbyte.cdk.db.jdbc.JdbcDatabase;
import io.airbyte.cdk.integrations.base.DestinationConfig;
import io.airbyte.cdk.integrations.base.JavaBaseConstants;
import io.airbyte.cdk.integrations.destination.NamingConventionTransformer;
import io.airbyte.cdk.integrations.standardtest.destination.DestinationAcceptanceTest;
import io.airbyte.cdk.integrations.standardtest.destination.argproviders.DataArgumentsProvider;
import io.airbyte.cdk.integrations.standardtest.destination.comparator.TestDataComparator;
import io.airbyte.commons.io.IOs;
import io.airbyte.commons.json.Jsons;
import io.airbyte.commons.resources.MoreResources;
import io.airbyte.commons.string.Strings;
import io.airbyte.configoss.StandardCheckConnectionOutput;
import io.airbyte.configoss.StandardCheckConnectionOutput.Status;
import io.airbyte.integrations.base.destination.typing_deduping.StreamId;
import io.airbyte.integrations.destination.snowflake.typing_deduping.SnowflakeSqlGenerator;
import io.airbyte.protocol.models.v0.AirbyteCatalog;
import io.airbyte.protocol.models.v0.AirbyteConnectionStatus;
import io.airbyte.protocol.models.v0.AirbyteMessage;
import io.airbyte.protocol.models.v0.CatalogHelpers;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
import java.nio.file.Path;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.TimeZone;
import java.util.stream.Collectors;
import javax.sql.DataSource;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ArgumentsSource;

@Disabled
public class SnowflakeInsertDestinationAcceptanceTest extends DestinationAcceptanceTest {

  private static final NamingConventionTransformer NAME_TRANSFORMER = new SnowflakeSQLNameTransformer();
  protected static final String NO_ACTIVE_WAREHOUSE_ERR_MSG =
      "No active warehouse selected in the current session.  Select an active warehouse with the 'use warehouse' command.";

  protected static final String NO_USER_PRIVILEGES_ERR_MSG =
      "Encountered Error with Snowflake Configuration: Current role does not have permissions on the target schema please verify your privileges";

  protected static final String IP_NOT_IN_WHITE_LIST_ERR_MSG = "is not allowed to access Snowflake. Contact your local security administrator";

  // this config is based on the static config, and it contains a random
  // schema name that is different for each test run
  private JsonNode config;
  private JdbcDatabase database;
  private DataSource dataSource;

  @BeforeEach
  public void setup() {
    DestinationConfig.initialize(getConfig());
  }

  @Override
  protected String getImageName() {
    return "airbyte/destination-snowflake:dev";
  }

  @Override
  protected JsonNode getConfig() {
    return config;
  }

  @Override
  protected TestDataComparator getTestDataComparator() {
    return new SnowflakeTestDataComparator();
  }

  @Override
  protected boolean supportBasicDataTypeTest() {
    return true;
  }

  @Override
  protected boolean supportArrayDataTypeTest() {
    return true;
  }

  @Override
  protected boolean supportObjectDataTypeTest() {
    return true;
  }

  protected boolean supportsInDestinationNormalization() {
    return true;
  }

  public JsonNode getStaticConfig() {
    final JsonNode insertConfig = Jsons.deserialize(IOs.readFile(Path.of("secrets/insert_config.json")));
    return insertConfig;
  }

  @Override
  protected JsonNode getFailCheckConfig() {
    final JsonNode invalidConfig = Jsons.clone(config);
    ((ObjectNode) invalidConfig.get("credentials")).put("password", "wrong password");
    return invalidConfig;
  }

  @Override
  protected List<JsonNode> retrieveRecords(final TestDestinationEnv env,
                                           final String streamName,
                                           final String namespace,
                                           final JsonNode streamSchema)
      throws Exception {
    final StreamId streamId = new SnowflakeSqlGenerator(0).buildStreamId(namespace, streamName, JavaBaseConstants.DEFAULT_AIRBYTE_INTERNAL_NAMESPACE);
    return retrieveRecordsFromTable(streamId.getRawName(), streamId.getRawNamespace())
        .stream()
        .map(r -> r.get(JavaBaseConstants.COLUMN_NAME_DATA))
        .collect(Collectors.toList());
  }

  @Override
  protected boolean implementsNamespaces() {
    return true;
  }

  @Override
  protected boolean supportNamespaceTest() {
    return true;
  }

  @Override
  protected Optional<NamingConventionTransformer> getNameTransformer() {
    return Optional.of(NAME_TRANSFORMER);
  }

  @Override
  protected List<JsonNode> retrieveNormalizedRecords(final TestDestinationEnv testEnv, final String streamName, final String namespace)
      throws Exception {
    final String tableName = NAME_TRANSFORMER.getIdentifier(streamName);
    final String schema = NAME_TRANSFORMER.getNamespace(namespace);
    return retrieveRecordsFromTable(tableName, schema);
  }

  private List<JsonNode> retrieveRecordsFromTable(final String tableName, final String schema) throws SQLException {
    final TimeZone timeZone = TimeZone.getTimeZone("UTC");
    TimeZone.setDefault(timeZone);

    return database.bufferedResultSetQuery(
        connection -> {
          try (final ResultSet tableInfo = connection.createStatement()
              .executeQuery(String.format("SHOW TABLES LIKE '%s' IN SCHEMA \"%s\";", tableName, schema))) {
            assertTrue(tableInfo.next());
            // check that we're creating permanent tables. DBT defaults to transient tables, which have
            // `TRANSIENT` as the value for the `kind` column.
            assertEquals("TABLE", tableInfo.getString("kind"));
            connection.createStatement().execute("ALTER SESSION SET TIMEZONE = 'UTC';");
            return connection.createStatement()
                .executeQuery(String.format("SELECT * FROM \"%s\".\"%s\" ORDER BY \"%s\" ASC;", schema, tableName,
                    JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT));
          }
        },
        new SnowflakeSourceOperations()::rowToJson);
  }

  // for each test we create a new schema in the database. run the test in there and then remove it.
  @Override
  protected void setup(final TestDestinationEnv testEnv, final HashSet<String> TEST_SCHEMAS) throws Exception {
    final String schemaName = Strings.addRandomSuffix("integration_test", "_", 5);
    final String createSchemaQuery = String.format("CREATE SCHEMA %s", schemaName);
    TEST_SCHEMAS.add(schemaName);

    this.config = Jsons.clone(getStaticConfig());
    ((ObjectNode) config).put("schema", schemaName);

    dataSource = SnowflakeDatabase.createDataSource(config, OssCloudEnvVarConsts.AIRBYTE_OSS);
    database = SnowflakeDatabase.getDatabase(dataSource);
    database.execute(createSchemaQuery);
  }

  @Override
  protected void tearDown(final TestDestinationEnv testEnv) throws Exception {
    getTestSchemas().add(config.get("schema").asText());
    for (final String schema : getTestSchemas()) {
      // we need to wrap namespaces in quotes, but that means we have to manually upcase them.
      // thanks, v1 destinations!
      // this probably doesn't actually work, because v1 destinations are mangling namespaces and names
      // but it's approximately correct and maybe works for some things.
      final String mangledSchema = schema.toUpperCase();
      final String dropSchemaQuery = String.format("DROP SCHEMA IF EXISTS \"%s\"", mangledSchema);
      database.execute(dropSchemaQuery);
    }

    DataSourceFactory.close(dataSource);
  }

  @Disabled("See README for why this test is disabled")
  @Test
  public void testCheckWithNoTextSchemaPermissionConnection() throws Exception {
    // Config to user (creds) that has no permission to schema
    final JsonNode config = Jsons.deserialize(IOs.readFile(
        Path.of("secrets/config_no_text_schema_permission.json")));

    final StandardCheckConnectionOutput standardCheckConnectionOutput = runCheck(config);

    assertEquals(Status.FAILED, standardCheckConnectionOutput.getStatus());
    assertThat(standardCheckConnectionOutput.getMessage()).contains(NO_USER_PRIVILEGES_ERR_MSG);
  }

  @Test
  public void testCheckIpNotInWhiteListConnection() throws Exception {
    // Config to user(creds) that has no warehouse assigned
    final JsonNode config = Jsons.deserialize(IOs.readFile(
        Path.of("secrets/insert_ip_not_in_whitelist_config.json")));

    final StandardCheckConnectionOutput standardCheckConnectionOutput = runCheck(config);

    assertEquals(Status.FAILED, standardCheckConnectionOutput.getStatus());
    assertThat(standardCheckConnectionOutput.getMessage()).contains(IP_NOT_IN_WHITE_LIST_ERR_MSG);
  }

  @Test
  public void testBackwardCompatibilityAfterAddingOauth() {
    final JsonNode deprecatedStyleConfig = Jsons.clone(config);
    final JsonNode password = deprecatedStyleConfig.get("credentials").get("password");

    ((ObjectNode) deprecatedStyleConfig).remove("credentials");
    ((ObjectNode) deprecatedStyleConfig).set("password", password);

    assertEquals(Status.SUCCEEDED, runCheckWithCatchedException(deprecatedStyleConfig));
  }

  @Test
  void testCheckWithKeyPairAuth() throws Exception {
    final JsonNode credentialsJsonString = Jsons.deserialize(IOs.readFile(Path.of("secrets/config_key_pair.json")));
    final AirbyteConnectionStatus check = new SnowflakeDestination(OssCloudEnvVarConsts.AIRBYTE_OSS).check(credentialsJsonString);
    assertEquals(AirbyteConnectionStatus.Status.SUCCEEDED, check.getStatus());
  }

  /**
   * This test is disabled because it is very slow, and should only be run manually for now.
   */
  @Disabled
  @ParameterizedTest
  @ArgumentsSource(DataArgumentsProvider.class)
  public void testSyncWithBillionRecords(final String messagesFilename, final String catalogFilename) throws Exception {
    final AirbyteCatalog catalog = Jsons.deserialize(MoreResources.readResource(catalogFilename), AirbyteCatalog.class);
    final ConfiguredAirbyteCatalog configuredCatalog = CatalogHelpers.toDefaultConfiguredCatalog(catalog);
    final List<AirbyteMessage> messages = MoreResources.readResource(messagesFilename).lines()
        .map(record -> Jsons.deserialize(record, AirbyteMessage.class)).toList();

    final List<AirbyteMessage> largeNumberRecords =
        Collections.nCopies(15000000, messages).stream().flatMap(List::stream).collect(Collectors.toList());

    final JsonNode config = getConfig();
    runSyncAndVerifyStateOutput(config, largeNumberRecords, configuredCatalog, false);
  }

}

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/kotlin/io/airbyte/integrations/destination/snowflake/SnowflakeInsertDestinationAcceptanceTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.destination.snowflake

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import io.airbyte.cdk.db.factory.DataSourceFactory.close
import io.airbyte.cdk.db.jdbc.JdbcDatabase
import io.airbyte.cdk.integrations.base.DestinationConfig
import io.airbyte.cdk.integrations.base.JavaBaseConstants
import io.airbyte.cdk.integrations.destination.NamingConventionTransformer
import io.airbyte.cdk.integrations.standardtest.destination.DestinationAcceptanceTest
import io.airbyte.cdk.integrations.standardtest.destination.argproviders.DataArgumentsProvider
import io.airbyte.cdk.integrations.standardtest.destination.comparator.TestDataComparator
import io.airbyte.commons.io.IOs.readFile
import io.airbyte.commons.json.Jsons
import io.airbyte.commons.json.Jsons.deserialize
import io.airbyte.commons.resources.MoreResources.readResource
import io.airbyte.commons.string.Strings.addRandomSuffix
import io.airbyte.configoss.StandardCheckConnectionOutput
import io.airbyte.integrations.destination.snowflake.typing_deduping.SnowflakeSqlGenerator
import io.airbyte.protocol.models.v0.AirbyteCatalog
import io.airbyte.protocol.models.v0.AirbyteConnectionStatus
import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.protocol.models.v0.CatalogHelpers
import java.nio.file.Path
import java.sql.Connection
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*
import java.util.stream.Collectors
import javax.sql.DataSource
import org.assertj.core.api.AssertionsForClassTypes
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ArgumentsSource

@Disabled
open class SnowflakeInsertDestinationAcceptanceTest : DestinationAcceptanceTest() {
    // this config is based on the static config, and it contains a random
    // schema name that is different for each test run
    open val staticConfig: JsonNode
        get() {
            val insertConfig = deserialize(readFile(Path.of("secrets/insert_config.json")))
            return insertConfig
        }
    private var config: JsonNode = Jsons.clone(staticConfig)
    private var dataSource: DataSource =
        SnowflakeDatabaseUtils.createDataSource(config, OssCloudEnvVarConsts.AIRBYTE_OSS)
    private var database: JdbcDatabase = SnowflakeDatabaseUtils.getDatabase(dataSource)

    @BeforeEach
    fun setup() {
        DestinationConfig.initialize(getConfig())
    }

    override val imageName: String
        get() = "airbyte/destination-snowflake:dev"

    override fun getConfig(): JsonNode {
        return config
    }

    override fun getTestDataComparator(): TestDataComparator {
        return SnowflakeTestDataComparator()
    }

    override fun supportBasicDataTypeTest(): Boolean {
        return true
    }

    override fun supportArrayDataTypeTest(): Boolean {
        return true
    }

    override fun supportObjectDataTypeTest(): Boolean {
        return true
    }

    override fun supportsInDestinationNormalization(): Boolean {
        return true
    }

    override fun getFailCheckConfig(): JsonNode? {
        val invalidConfig: JsonNode = Jsons.clone<JsonNode>(config)
        (invalidConfig["credentials"] as ObjectNode).put("password", "wrong password")
        return invalidConfig
    }

    @Throws(Exception::class)
    override fun retrieveRecords(
        testEnv: TestDestinationEnv?,
        streamName: String,
        namespace: String,
        streamSchema: JsonNode
    ): List<JsonNode> {
        val streamId =
            SnowflakeSqlGenerator(0)
                .buildStreamId(
                    namespace,
                    streamName,
                    JavaBaseConstants.DEFAULT_AIRBYTE_INTERNAL_NAMESPACE
                )
        return retrieveRecordsFromTable(streamId.rawName, streamId.rawNamespace)
            .stream()
            .map { r: JsonNode -> r.get(JavaBaseConstants.COLUMN_NAME_DATA) }
            .collect(Collectors.toList<JsonNode>())
    }

    override fun implementsNamespaces(): Boolean {
        return true
    }

    override fun supportNamespaceTest(): Boolean {
        return true
    }

    override fun getNameTransformer(): Optional<NamingConventionTransformer> {
        return Optional.of(NAME_TRANSFORMER)
    }

    @SuppressFBWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE")
    @Throws(Exception::class)
    override fun retrieveNormalizedRecords(
        testEnv: TestDestinationEnv?,
        streamName: String?,
        namespace: String?
    ): List<JsonNode> {
        val tableName = NAME_TRANSFORMER.getIdentifier(streamName!!)
        val schema = NAME_TRANSFORMER.getNamespace(namespace!!)
        return retrieveRecordsFromTable(tableName, schema)
    }

    @Throws(SQLException::class)
    private fun retrieveRecordsFromTable(tableName: String, schema: String): List<JsonNode> {
        val timeZone = TimeZone.getTimeZone("UTC")
        TimeZone.setDefault(timeZone)

        return database.bufferedResultSetQuery<JsonNode>(
            { connection: Connection ->
                connection
                    .createStatement()
                    .executeQuery(
                        String.format("SHOW TABLES LIKE '%s' IN SCHEMA \"%s\";", tableName, schema)
                    )
                    .use { tableInfo ->
                        Assertions.assertTrue(tableInfo.next())
                        // check that we're creating permanent tables. DBT defaults to transient
                        // tables, which have
                        // `TRANSIENT` as the value for the `kind` column.
                        Assertions.assertEquals("TABLE", tableInfo.getString("kind"))
                        connection.createStatement().execute("ALTER SESSION SET TIMEZONE = 'UTC';")
                        return@bufferedResultSetQuery connection
                            .createStatement()
                            .executeQuery(
                                String.format(
                                    "SELECT * FROM \"%s\".\"%s\" ORDER BY \"%s\" ASC;",
                                    schema,
                                    tableName,
                                    JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT
                                )
                            )
                    }
            },
            { queryResult: ResultSet -> SnowflakeSourceOperations().rowToJson(queryResult) }
        )
    }

    // for each test we create a new schema in the database. run the test in there and then remove
    // it.
    @Throws(Exception::class)
    override fun setup(testEnv: TestDestinationEnv, TEST_SCHEMAS: HashSet<String>) {
        val schemaName = addRandomSuffix("integration_test", "_", 5)
        val createSchemaQuery = String.format("CREATE SCHEMA %s", schemaName)
        TEST_SCHEMAS.add(schemaName)

        this.config = Jsons.clone<JsonNode>(staticConfig)
        (config as ObjectNode?)!!.put("schema", schemaName)

        dataSource =
            SnowflakeDatabaseUtils.createDataSource(config, OssCloudEnvVarConsts.AIRBYTE_OSS)
        database = SnowflakeDatabaseUtils.getDatabase(dataSource)
        database.execute(createSchemaQuery)
    }

    @Throws(Exception::class)
    override fun tearDown(testEnv: TestDestinationEnv) {
        testSchemas.add(config["schema"].asText())
        for (schema in testSchemas) {
            // we need to wrap namespaces in quotes, but that means we have to manually upcase them.
            // thanks, v1 destinations!
            // this probably doesn't actually work, because v1 destinations are mangling namespaces
            // and names
            // but it's approximately correct and maybe works for some things.
            val mangledSchema = schema.uppercase(Locale.getDefault())
            val dropSchemaQuery = String.format("DROP SCHEMA IF EXISTS \"%s\"", mangledSchema)
            database.execute(dropSchemaQuery)
        }

        close(dataSource)
    }

    @Disabled("See README for why this test is disabled")
    @Test
    @Throws(Exception::class)
    fun testCheckWithNoTextSchemaPermissionConnection() {
        // Config to user (creds) that has no permission to schema
        val config = deserialize(readFile(Path.of("secrets/config_no_text_schema_permission.json")))

        val standardCheckConnectionOutput = runCheck(config)

        Assertions.assertEquals(
            StandardCheckConnectionOutput.Status.FAILED,
            standardCheckConnectionOutput.status
        )
        AssertionsForClassTypes.assertThat(standardCheckConnectionOutput.message)
            .contains(NO_USER_PRIVILEGES_ERR_MSG)
    }

    @Test
    @Throws(Exception::class)
    fun testCheckIpNotInWhiteListConnection() {
        // Config to user(creds) that has no warehouse assigned
        val config =
            deserialize(readFile(Path.of("secrets/insert_ip_not_in_whitelist_config.json")))

        val standardCheckConnectionOutput = runCheck(config)

        Assertions.assertEquals(
            StandardCheckConnectionOutput.Status.FAILED,
            standardCheckConnectionOutput.status
        )
        AssertionsForClassTypes.assertThat(standardCheckConnectionOutput.message)
            .contains(IP_NOT_IN_WHITE_LIST_ERR_MSG)
    }

    @Test
    fun testBackwardCompatibilityAfterAddingOauth() {
        val deprecatedStyleConfig: JsonNode = Jsons.clone<JsonNode>(config)
        val password = deprecatedStyleConfig["credentials"]["password"]

        (deprecatedStyleConfig as ObjectNode).remove("credentials")
        deprecatedStyleConfig.set<JsonNode>("password", password)

        Assertions.assertEquals(
            StandardCheckConnectionOutput.Status.SUCCEEDED,
            runCheckWithCatchedException(deprecatedStyleConfig)
        )
    }

    @Test
    @Throws(Exception::class)
    fun testCheckWithKeyPairAuth() {
        val credentialsJsonString = deserialize(readFile(Path.of("secrets/config_key_pair.json")))
        val check =
            SnowflakeDestination(OssCloudEnvVarConsts.AIRBYTE_OSS).check(credentialsJsonString)
        Assertions.assertEquals(AirbyteConnectionStatus.Status.SUCCEEDED, check!!.status)
    }

    /** This test is disabled because it is very slow, and should only be run manually for now. */
    @Disabled
    @ParameterizedTest
    @ArgumentsSource(DataArgumentsProvider::class)
    @Throws(Exception::class)
    fun testSyncWithBillionRecords(messagesFilename: String, catalogFilename: String) {
        val catalog = deserialize(readResource(catalogFilename), AirbyteCatalog::class.java)
        val configuredCatalog = CatalogHelpers.toDefaultConfiguredCatalog(catalog)
        val messages: List<AirbyteMessage> =
            readResource(messagesFilename)
                .lines()
                .map { record: String -> deserialize(record, AirbyteMessage::class.java) }
                .toList()

        val largeNumberRecords =
            Collections.nCopies(15000000, messages)
                .stream()
                .flatMap { obj: List<AirbyteMessage> -> obj.stream() }
                .collect(Collectors.toList())

        val config = getConfig()
        runSyncAndVerifyStateOutput(config, largeNumberRecords, configuredCatalog, false)
    }

    companion object {
        private val NAME_TRANSFORMER: NamingConventionTransformer = SnowflakeSQLNameTransformer()
        @JvmStatic
        protected val NO_ACTIVE_WAREHOUSE_ERR_MSG: String =
            "No active warehouse selected in the current session.  Select an active warehouse with the 'use warehouse' command."

        @JvmStatic
        protected val NO_USER_PRIVILEGES_ERR_MSG: String =
            "Encountered Error with Snowflake Configuration: Current role does not have permissions on the target schema please verify your privileges"

        @JvmStatic
        protected val IP_NOT_IN_WHITE_LIST_ERR_MSG: String =
            "is not allowed to access Snowflake. Contact your local security administrator"
    }
}
